Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90217

Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

Commit:
Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors.

Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

[ghstack-poisoned]

ngimel reviewed (approved):
Hope it sticks this time (which unfortunately is my comment on too many diffs)

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -74,25 +87,38 @@ static void accumulate(
+    buffer[pos] = old_var + var;
+  } else if (
+      // ATen doesn't route sparse additions correctly...
+      old_var.is_sparse() || old_var.is_sparse_csr()) {
```
cc @cpuhrsch what is the right way (and future proof way) to check if a Tensor is any sparse format?

cpuhrsch replied:
@albanD - a future proof catch-all is work in progress. Right now it's best to access the layout and compare to the various sparse formats.

`self.layout() == kSparseBsc`, `self.layout() == kSparseBsr`, `self.layout() == kSparseCsc`, `self.layout() == kSparseCsr`, `self.layout() == kSparse` (see [Layout.h](https://github.com/pytorch/pytorch/blob/master/c10/core/Layout.h#L11-L16)).

cc @amjames for visibility. I created a new entry in the roadmap for this so we can track it.

Commit:
Update on "Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors."


Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

Commit:
Update on "Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors."


Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

Commit:
Update on "Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors."


Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

Commit:
Update on "Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors."


Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

robieta commented:
@pytorchbot merge -f "unit test failures are preexisting"
